File: Cpp/Maximum Subarray Difference.cpp
#include <vector>

class Solution {
public:
    /**
     * @param nums: A list of integers
     * @return: An integer indicate the value of maximum difference between two
     *          Subarrays
     */
    int abs(int s) {
        if (s < 0) return -s;
        else return s;
    }
    int maxDiffSubArrays(vector<int> nums) {
        // write your code here
        int len = nums.size();
        vector<int> leftmin(len), leftmax(len), rightmin(len), rightmax(len);
        int res1 = nums[0], cnt1 = 0, res2 = nums[0], cnt2 = 0;
        for (int i = 0; i < len; i ++) {
            cnt1 = max(cnt1 + nums[i], nums[i]);
            cnt2 = min(cnt2 + nums[i], nums[i]);
            res1 = max(cnt1, res1);
            res2 = min(cnt2, res2);
            leftmax[i] = res1;
            leftmin[i] = res2;
        }
        res1 = nums[len-1], cnt1 = 0, res2 = nums[len-1], cnt2 = 0;
        for (int i = len - 1; i >= 0; i --) {
            cnt1 = max(cnt1 + nums[i], nums[i]);
            cnt2 = min(cnt2 + nums[i], nums[i]);
            res1 = max(cnt1, res1);
            res2 = min(cnt2, res2);
            rightmax[i] = res1;
            rightmin[i] = res2;
        }
        int maxdiffsub = 0;
        for (int i = 0; i < len - 1; i ++) {
            maxdiffsub = max(maxdiffsub, max(abs(leftmax[i] - rightmin[i+1]), abs(rightmax[i+1] - leftmin[i])));
        }
        return maxdiffsub;
    }
};

File: Java/Maximum-Subarray-Difference.java
public class Solution {
    /**
     * @param nums: A list of integers
     * @return: An integer indicate the value of maximum difference between two
     *          Subarrays
     */
    public int maxDiffSubArrays(ArrayList<Integer> nums) {
        // write your code
        int len = nums.size();
        int []leftmax = new int[len];
        int []leftmin = new int[len];
        int cur1 = -1, temp1 = Integer.MIN_VALUE, cur2 = 1, temp2 = Integer.MAX_VALUE;
        for (int i = 0; i < len; i ++) {
            cur1 = Math.max(nums.get(i), nums.get(i) + cur1);
            temp1 = Math.max(temp1, cur1);
            leftmax[i] = temp1;
            cur2 = Math.min(nums.get(i), nums.get(i) + cur2);
            temp2 = Math.min(temp2, cur2);
            leftmin[i] = temp2;
        }
        int res = 0;
        cur1 = -1;
        cur2 = 1;
        temp1 = Integer.MIN_VALUE;
        temp2 = Integer.MAX_VALUE;
        for (int i = len - 1; i > 0; i --) {
            cur1 = Math.max(nums.get(i), nums.get(i) + cur1);
            temp1 = Math.max(temp1, cur1);
            res = Math.max(res, Math.abs(temp1 - leftmin[i - 1]));
            cur2 = Math.min(nums.get(i), nums.get(i) + cur2);
            temp2 = Math.min(temp2, cur2);
            res = Math.max(res, Math.abs(temp2 - leftmax[i - 1]));
        }
        return res;
    }
}


